### DISCUSSION TITLE
Multiple layer, multiple shape collision filtering as described in the docs

### DISCUSSION THREAD

**Discussion by z-adams** [jrouwe/JoltPhysics#1331] [General]
Apologies if this is explained clearly somewhere, but I couldn't find it and would like to add it to architecture.md if I can figure it out.

The manual ([§ Collision Filtering](https://jrouwe.github.io/JoltPhysics/index.html#collision-filtering)) describes a situation where collision filtering is used to give an object two collision/intersection behaviors: a regular geometric collider for common rigid body collisions (in layer MOVING), and a second high detail collider (e.g. a mesh collider) used for raycasting (in layer BULLET). This is precisely the behavior I am looking for -- but I'm not sure how the API wants this done.

Naively I was going to just create a separate body for each, making the high-res body kinematic and applying a fixed constraint to join it to its low-res dynamic parent, but this feels somewhat wasteful. This is for an application that supports large world multiplayer that may contain very large numbers of bodies at once, so doubling the number of bodies and needing to keep track of them + a constraint feels like a potentially great waste of cycles and memory (both on my side and within Jolt). Is there a way this relationship between the low and high-res collision shapes can be handled at a lower level to avoid this duplication, while retaining the collision filtering/broadphase capabilities? Thanks.

**Comment by jrouwe** [comment]
It is indeed impossible for a body to reside in 2 layers. If you want a high-res kinematic body following a low-res dynamic body, you'll need to create 2 Bodies. Using a fixed constraint in this case will not work as a kinematic object cannot be moved by a constraint (a kinematic object has infinite mass so cannot be moved by a dynamic object). You'll need to update the physics system and afterwards loop over your dynamic bodies and call `BodyInterface::MoveKinematic` to make the high-res kinematic body follow the low-res dynamic body. If your high-res kinematic bodies cannot interact with any dynamic objects, you can also just call `BodyInterface::SetPositionAndRotation` and instantly update the position. This might feel wasteful, but is exactly what Jolt would need to do if it were to implement this functionality.

**Reply by jrouwe** [reply]
I added this to the docs now as well: https://jrouwe.github.io/JoltPhysics/#level-of-detail

**Reply by z-adams** [reply]
Ok understood, thanks for the information. So there's no way to attach both shapes and then use a broadphase event handler to exclude one from the narrow phase? I'm not yet familiar enough with the architecture to reason deeply about this stuff.

You might also want to clarify the language in this line:

> BULLET - Layer for high detail collision bodies that we attach to regular dynamic bodies. These are not used for simulation but we want extra precision when we shoot with bullets.

since the "attach" verbiage is where I inferred a constraint was the way to go here. Might be better stated as something like "Layer for high detail collision bodies that we co-locate with regular dynamic bodies. These are kinematic bodies that are not used for simulation but are moved to follow the dynamic bodies and provide more precise geometry for ray tests to simulate shooting bullets.

**Reply by jrouwe** [reply]
Done.

**Reply by jrouwe** [reply]
> So there's no way to attach both shapes and then use a broadphase event handler to exclude one from the narrow phase?

For collision detection you can use a `ShapeFilter` to exclude sub-shapes, see [this example](https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/General/ShapeFilterTest.cpp). Unfortunately this construct doesn't exist for the simulation at the moment, so you cannot exclude a sub-shape there. It is something that I want to support in the future though.

**Reply by LPGhatguy** [reply]
With the new `SimShapeFilter` interface, you can now filter collisions at the shape level in the simulation. I opened a PR to update the doc link in the "level of detail" section to point to the new type.
